core/src/main/java/org/elasticsearch/index/query/GeoBoundingBoxQueryBuilder.java

```diff
@@ -97,94 +155,286 @@ public GeoBoundingBoxQueryBuilder bottomRight(String geohash) {
         return this;
     }
 
+    /**
+     * For BWC: Parse type from type name.
+     * */
+    public GeoBoundingBoxQueryBuilder type(String type) {
+        this.type = GeoExecType.fromString(type);
+        return this;
+    }
+    /** Returns the execution type of the geo bounding box.*/
+    public GeoExecType type() {
+        return type;
+    }
+
+    /** Returns the name of the field to base the bounding box computation on. */
+    public String fieldName() {
+        return this.fieldName;
+    }
+
     @Override
-    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
-        // check values
-        if(Double.isNaN(box[TOP])) {
-            throw new IllegalArgumentException("geo_bounding_box requires top latitude to be set");
-        } else if(Double.isNaN(box[BOTTOM])) {
-            throw new IllegalArgumentException("geo_bounding_box requires bottom latitude to be set");
-        } else if(Double.isNaN(box[RIGHT])) {
-            throw new IllegalArgumentException("geo_bounding_box requires right longitude to be set");
-        } else if(Double.isNaN(box[LEFT])) {
-            throw new IllegalArgumentException("geo_bounding_box requires left longitude to be set");
+    public QueryValidationException validate() {
+        QueryValidationException validationException = null;
+        if (!ignoreMalformed) {
+            if (Numbers.isValidDouble(topLeft.getLat()) == false) {
+                validationException = addValidationError("top latitude is invalid: " + topLeft.getLat(), validationException);
+            }
+            if (Numbers.isValidDouble(topLeft.getLon()) == false) {
+                validationException = addValidationError("left longitude is invalid: " + topLeft.getLon(), validationException);
+            }
+            if (Numbers.isValidDouble(bottomRight.getLat()) == false) {
+                validationException = addValidationError("bottom latitude is invalid: " + bottomRight.getLat(), validationException);
+            }
+            if (Numbers.isValidDouble(bottomRight.getLon()) == false) {
+                validationException = addValidationError("right longitude is invalid: " + bottomRight.getLon(), validationException);
+            }
+
+            if (validationException == null) {
+                // all corners are valid after above checks - make sure they are in the right relation
+                if (topLeft.getLat() < bottomRight.getLat()) {
+                    validationException = addValidationError("top is below bottom corner: " +
+                            topLeft.getLat() + " vs. " + bottomRight.getLat(), validationException);
+                }
+
+                // we do not check longitudes as the query generation code can deal with flipped left/right values
+            }
         }
+        return validationException;
+    }
 
-        builder.startObject(NAME);
+    QueryValidationException checkLatLon(boolean indexCreatedBeforeV2_0) {
```

Comment:
https://github.com/elastic/elasticsearch/blob/b2ba3847f73ace98d73ea63e82b44cf5be326add/core/src/main/java/org/elasticsearch/index/query/GeoBoundingBoxQueryParser.java#L161 is where we did it before. The check for indexcreation version is there to accommodate older percolation queries. 

The reason why I think checking lat/lon values makes sense at all: It's the simplest thing we can do to help users detect implementation errors where they've swapped lat/lon values. I've made this particular mistake myself too often in the past, I guess.

Leaving in for now, no particularly strong opinion here though.